images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/__init__.py

```diff
@@ -15,44 +18,94 @@
 from .actions import ActionHandler
 from .gpios import GpioManager
 
+from x1plus.utils import module_importer
+
 logger = logging.getLogger(__name__)
 
 class X1PlusDaemon:
+
     @classmethod
     async def create(cls):
         # must be a classmethod, since __init__ cannot be async
         self = X1PlusDaemon()
 
         logger.info("creating x1plusd services")
+        
+        # Required modules
         self.router = await get_dbus_router()
         self.settings = SettingsService(router=self.router)
+
         self.mqtt = MQTTClient(daemon=self)
-        self.ota = OTAService(router=self.router, daemon=self)
+        self.ota = OTAService(daemon=self)
         self.ssh = SSHService(daemon=self)
-        self.httpd = HTTPService(router=self.router, daemon=self)
-        self.sensors = SensorsService(router=self.router, daemon=self)
-        self.gpios = GpioManager(daemon=self)
-        self.expansion = ExpansionManager(router=self.router, daemon=self)
+        self.httpd = HTTPService(daemon=self)
+        self.sensors = SensorsService(daemon=self)
         self.mcproto = MCProtoParser(daemon=self)
-        self.actions = ActionHandler(router=self.router, daemon=self)
-        if not self.settings.get("polar_cloud", False):
-            self.polar_cloud = None
-        else:
-            from .polar_cloud import PolarPrintService
-            self.polar_cloud = PolarPrintService(daemon=self)
+        self.expansion = ExpansionManager(daemon=self)
+        self.gpios = GpioManager(daemon=self)
+        self.actions = ActionHandler(daemon=self)
         
+        self.custom_modules = {}
+        """
+        { "name": ClassInstance }
+        """
+
+        self.watched_keys = []
+        self.MODULES = []
+
+        BASE_MODULE_DIR = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), "modules"))
+        self.load_modules(BASE_MODULE_DIR)
+
+        SDCARD_MODULE_DIR = "/mnt/sdcard/x1plus/modules"
+        if not os.path.isdir(SDCARD_MODULE_DIR):
+            os.makedirs(SDCARD_MODULE_DIR)
+
+        # Custom modules are loaded after all core modules
+        # Full import paths will work if a dependency exists
+        self.load_modules(SDCARD_MODULE_DIR)
+
         return self
 
+    def load_modules(self, module_directory: str):
+        self.MODULES.extend(module_importer(module_directory, include_subdirs=True))
```

Comment:
so I think we don't actually want to do the actual `import` until we've agreed (from the settings subsystem) that it's ok to import -- we could haul in a bunch of dependencies (for people trying to run a lean-and-mean x1plusd), and the act of `import`ing something is sufficient to cause code to execute!  so I think `MODULES` ought really be like `modules_available`.